### PR TITLE
feat: Check whether the data validator is registered in Panacea

### PR DESCRIPTION
**Pull request by inchori**
- Duplicate Check for Registered Data Validator using `QueryDataValidator` in Panacea.
- It checks duplication when the `dataval start`.

### CODE REVIEW

File: panacea/grpc_client.go [hunk 1]
```diff
@@ -123,6 +123,20 @@ func (c *GrpcClient) GetChainId() (string, error) {
...
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+
+	res, err := client.DataValidator(ctx, &datapooltypes.QueryDataValidatorRequest{Address: address})
+	if err != nil {
+		return datapooltypes.DataValidator{}, err
+	}
+
+	return *res.GetDataValidator(), err
```
Comment (gyuguen) [inline panacea/grpc_client.go:137]: I think it's better to return nil for errors.
Of course, `err` is always nil, but it seems better to declare it expilicitly.
```suggestion
	return *res.GetDataValidator(), nil
```

Comment (inchori) [reply panacea/grpc_client.go:137]: Oh, My mistake. I'll fix it.
> but it seems better to declare it expilicitly
And I'll add `fmt.Errorf` too. Thanks

File: panacea/grpc_client.go [hunk 2]
```diff
@@ -123,6 +123,20 @@ func (c *GrpcClient) GetChainId() (string, error) {
 	return response.GetBlock().GetHeader().ChainID, nil
 }
 
+func (c *GrpcClient) GetRegisteredDataValidator(address string) (datapooltypes.DataValidator, error) {
+	client := datapooltypes.NewQueryClient(c.conn)
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+
+	res, err := client.DataValidator(ctx, &datapooltypes.QueryDataValidatorRequest{Address: address})
```
Comment (gyuguen) [inline panacea/grpc_client.go:132]: This is something I haven't been able to review before.
If the `DataValidator` is not reigstered, it does not respond with an error and returns an empty value.
How about responding with an error if `DataValidator` is not registered?

**Comment by youngjoon-lee** [comment]
When you merge this PR to the `main`, please rename the title from `Duplicate Check` to `Duplication Check`. The duplicate check means the double-checking which checks something twice for safety.

**Comment by inchori** [comment]
Modify Cosmwasm `.so` file in Dockerfile based on panacea-core [Dockerfile](https://github.com/medibloc/panacea-core/blob/master/Dockerfile#:~:text=ADD%20https%3A//github,a%20%7C%20grep%203f5de8df9c6b606b4211f90edd681c84b0ecd870fdbf50678b6d9afd783a571c).

File: config/config.go [hunk 3]
```diff
@@ -5,17 +5,17 @@ import "fmt"
...
+	Panacea PanaceaConfig `mapstructure:"panacea"`
+	AWSS3   AWSS3Config   `mapstructure:"aws-s3"`
+	Enclave EnclaveConfig `mapstructure:"enclave"`
 }
 
 type BaseConfig struct {
 	LogLevel              string `mapstructure:"log-level"`
 	ValidatorMnemonic     string `mapstructure:"validator-mnemonic"`
 	DataEncryptionKeyFile string `mapstructure:"data-encryption-key-file"`
+	EndPoint              string `mapstructure:"endpoint"`
```
Comment (youngjoon-lee) [inline config/config.go:18]: I think it would be good to put this value into the `HTTPConfig` as Tendermint does. Also, please rename it to `Endpoint`.

Comment (youngjoon-lee) [reply config/config.go:18]: Or, `ExternalEndpoint`, whatever you prefer.

File: panacea/grpc_client.go [hunk 4]
```diff
@@ -131,96 +147,106 @@ func (c *GrpcClient) RegisterDataValidator(endpoint string, validatorAcc *Valida
...
-		ctx,
-		&txtypes.BroadcastTxRequest{
-			Mode:    txtypes.BroadcastMode_BROADCAST_MODE_BLOCK,
-			TxBytes: txBytes,
-		},
-	)
-	if err != nil {
-		return nil
-	}
+	getRegisteredDataValidator, _ := c.GetRegisteredDataValidator(address)
```
Comment (youngjoon-lee) [inline panacea/grpc_client.go:151]: Please handle an error

Comment (youngjoon-lee) [reply panacea/grpc_client.go:151]: Also, please rename the return value properly.

File: panacea/grpc_client.go [hunk 5]
```diff
@@ -131,96 +147,106 @@ func (c *GrpcClient) RegisterDataValidator(endpoint string, validatorAcc *Valida
...
+
+	var dataValidator datapooltypes.DataValidator
+
+	if getRegisteredDataValidator == dataValidator {
```
Comment (youngjoon-lee) [inline panacea/grpc_client.go:155]: When you write `if` statements, it's generally better to handle a shorter case. In this case, you can write:
```go
registeredDataVal, err := c.GetRegisteredDataValidator(address)
if err != nil {
    ...
}

if registeredDataVal != nil {
    // DO NOT return an error if you aren't gonna handle it proper outside of this function.
    // In this case, it's okay to return nil. Then, please put this info to the function comment.
    return nil
}

// Here, you register the dataval to Panacea
```

To do this, you need to change the function signature of `GetRegisteredDataValidator()` to return `*datapooltypes.DataValidator`, not `datapooltypes.DataValidator`.

File: config/config.go (hunk 3 above)
Comment (youngjoon-lee) [reply config/config.go:18]: I just saw comments in the other PR: https://github.com/medibloc/panacea-data-market-validator/pull/65#discussion_r840339468.
I think it's better to minimize the `BaseConfig`. The `Endpoint` may be tightly related to HTTP settings.

Comment (inchori) [reply config/config.go:18]: Ok I'll change it. URL is actually start with http, so I agree with your comments.

**Comment by inchori** [comment]
This branch was rebased from main branch.

Close #65 as we discussed.

Restore the `docker-compose.yml` because registering the data validator will be performed manually.

File: panacea/grpc_client.go [hunk 6]
```diff
@@ -41,16 +43,30 @@ func makeInterfaceRegistry() sdk.InterfaceRegistry {
...
 
-func (c *GrpcClient) Close() {
+func (c *GrpcClient) Close() error {
 	log.Info("closing Panacea gRPC connection")
-	c.conn.Close()
+	err := c.conn.Close()
+	if err != nil {
+		return err
+	}
+	return nil
```
Comment (youngjoon-lee) [inline panacea/grpc_client.go:56]: In this case, you can just `return c.conn.Close()`.

File: panacea/grpc_client.go [hunk 7]
```diff
@@ -86,6 +102,35 @@ func (c *GrpcClient) GetDeal(id string) (datadealtypes.Deal, error) {
...
+
+	response, err := client.GetLatestBlock(ctx, &tmservice.GetLatestBlockRequest{})
+	if err != nil {
+		return "", err
+	}
+
+	return response.GetBlock().GetHeader().ChainID, nil
+}
+
+// GetRegisteredDataValidator gets registered data validator for checking duplication from blockchain
```
Comment (youngjoon-lee) [inline panacea/grpc_client.go:119]:
```suggestion
// GetRegisteredDataValidator gets registered data validator
```

File: server/service/service.go [hunk 8]
```diff
@@ -34,6 +33,11 @@ func New(conf *config.Config) (*Service, error) {
 		return nil, fmt.Errorf("failed to create PanaceaGRPCClient: %w", err)
 	}
 
+	_, err = panaceaClient.GetRegisteredDataValidator(validatorAccount.GetAddress())
+	if err != nil {
+		return nil, err
+	}
```
Comment (youngjoon-lee) [inline server/service/service.go:39]:
```suggestion
	registeredDataVal, err = panaceaClient.GetRegisteredDataValidator(validatorAccount.GetAddress())
	if err != nil {
		return nil, err
	} else if registeredDataVal == nil {
	        return nil, fmt.Errorf("this data validator is not registered in Panacea yet")
	}
```

Comment (youngjoon-lee) [reply server/service/service.go:39]: Plus, please change the title of this PR: `Check whether the data validator is registered in Panacea`.

Comment (inchori) [reply server/service/service.go:39]: Ok, I'll rename it

File: panacea/grpc_client.go [hunk 9]
```diff
@@ -86,6 +98,35 @@ func (c *GrpcClient) GetDeal(id string) (datadealtypes.Deal, error) {
...
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+
+	response, err := client.GetLatestBlock(ctx, &tmservice.GetLatestBlockRequest{})
+	if err != nil {
+		return "", err
+	}
+
+	return response.GetBlock().GetHeader().ChainID, nil
+}
```
Comment (youngjoon-lee) [inline panacea/grpc_client.go:113]: It seems that this function is not used.

**Review by youngjoon-lee — APPROVED** [review]
Thank you.

If you want me to re-review this PR after you push some additional commits, please click the `re-request review` button, so that I can get notified.
https://egghead.io/lessons/github-re-request-a-pr-review-on-github

File: server/service/service.go [hunk 10]
```diff
@@ -34,6 +33,13 @@ func New(conf *config.Config) (*Service, error) {
 		return nil, fmt.Errorf("failed to create PanaceaGRPCClient: %w", err)
 	}
 
+	registeredDataValidator, err := panaceaClient.GetRegisteredDataValidator(validatorAccount.GetAddress())
+	if err != nil {
+		return nil, err
+	} else if registeredDataValidator == nil {
+		return nil, fmt.Errorf("this data validator is not registered in Panacea yet")
+	}
```
Comment (gyuguen) [inline server/service/service.go:41]: My guess is that there is no case where `registeredDataValidator` is empty without an error.
There is a way to delete this part.
```go
else if registeredDataValidator == nil {
	return nil, fmt.Errorf("this data validator is not registered in Panacea yet")
}
```

Or, for a more accurate response, check for duplicate errors and respond to this message.(`this data validator is not registered in Panacea yet`)

Comment (inchori) [reply server/service/service.go:41]: I agree with your comment. 
Plus, Panacea will return error message but it is not suitable for data validator.
I'll change it.

Comment (inchori) [reply server/service/service.go:41]:
```go
_, err = panaceaClient.GetRegisteredDataValidator(validatorAccount.GetAddress())
if err != nil {
      if strings.HasSuffix(err.Error(), datapooltypes.ErrDataValidatorNotFound.Error()) {
           return nil, fmt.Errorf("this data validator is not registered in Panacea yet")
      }
      return nil, err
}
```
How about change like this?

Comment (gyuguen) [reply server/service/service.go:41]: Yes, It seems good.

**Review by gyuguen — APPROVED** [review]
LGTM 👍 
I think it would be nice to add E2E tests later.

**Pull request merged**